starlite/security/base.py

```diff
@@ -0,0 +1,166 @@
+from abc import ABC, abstractmethod
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Callable,
+    Dict,
+    Generic,
+    Iterable,
+    List,
+    Optional,
+    Type,
+    TypeVar,
+    Union,
+)
+
+from pydantic import BaseConfig, validator
+from pydantic.generics import GenericModel
+
+from starlite.connection import ASGIConnection
+from starlite.datastructures import Provide
+from starlite.middleware.authentication import AbstractAuthenticationMiddleware
+from starlite.types import ControllerRouterHandler, Guard, Scopes, SyncOrAsyncUnion
+from starlite.utils.sync import AsyncCallable
+
+if TYPE_CHECKING:
+    from pydantic_openapi_schema.v3_1_0 import Components, SecurityRequirement
+
+    from starlite.config import AppConfig
+    from starlite.middleware.base import DefineMiddleware
+
+
+UserType = TypeVar("UserType")
+AuthType = TypeVar("AuthType")
+
+
+class AbstractSecurityConfig(ABC, Generic[UserType, AuthType], GenericModel):
```

Comment:
I'm thinking if this shouldn't be the config for a plugin. Then you wouldn't need to manually set up the middleware and `on_init` hooks.

Reply:
we would need to set a plugin instead, its exactly the same amount of work. These are not plugins though - and they wont implement most of the methods specified by the plugin protocol, so there is no point in that.

Reply:
I don't agree that it's the same amount of work. Passing in the middleware and `on_init` handler vs simply passing in a plugin. The latter feels more "works out of the box". But I'm fine with not having it a plugin as well.

Reply:
but you dont pass the middleware, you just pass the `on_init` handler. See all the examples and tests